### DISCUSSION TITLE
Whether Milvus supports user-defined indexes and reserves user-defined index interfaces

### DISCUSSION THREAD

**Discussion by liuyichen2013** [milvus-io/milvus#18564] [Q&A and General discussion]
Whether Milvus supports user-defined indexes and reserves user-defined index interfaces，thanks

**Comment by yhmo** [comment]
Milvus doesn't support user-defined index. And no such interfaces provided.
To define an index, we need to provide several definitions:
1. the cluster method --- how to cluster vectors(for example: KNN, Graph)
2. the quantization method -- how to convert a vector into bytes(for example: PQ, SQ8)
3. the metric formula --- how to calculate the distance of two vectors

It is difficult to define these things via an interface. The next difficulty is how the index engine can recognize these cluster method, quantization method and metric formula, maybe JIT(just-in-time compilation) is a possible technology to make it.
